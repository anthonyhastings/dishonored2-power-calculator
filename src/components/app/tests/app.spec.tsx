import renderer from 'react-test-renderer';
import { App } from '../';
import { useAppDispatch } from '@/store-hooks';

jest.mock('react-router-dom', () => ({
  Routes: 'MockRoutes',
  Route: function MockRoute(props: { element: React.ReactNode }) {
    const { element: Element, ...otherProps } = props;
    return <span {...otherProps}>{Element}</span>;
  },
  Link: 'MockLink',
}));

jest.mock('selectors', () => ({
  hasInitialDataFailedSelector: jest.fn().mockReturnValue(false),
  isInitialDataIncompleteSelector: jest.fn().mockReturnValue(true),
}));

jest.mock('store-hooks', () => ({
  useAppSelector: (fn: () => void) => fn(),
  useAppDispatch: jest.fn(),
}));

jest.mock('slices/characters', () => ({
  fetchCharacters: () => 'MockFetchCharactersAction',
}));

jest.mock('slices/powers', () => ({
  fetchPowers: () => 'MockFetchPowersAction',
}));

jest.mock('components/loader', () => ({
  Loader: 'MockLoader',
}));

jest.mock(
  'pages/character-selection',
  () =>
    function MockCharacterSelection() {
      return <p>Mock Character Selection</p>;
    }
);

jest.mock(
  'pages/page-not-found',
  () =>
    function MockPageNotFound() {
      return <p>Page Not Found</p>;
    }
);

jest.mock(
  'pages/power-selection',
  () =>
    function MockPowerSelection() {
      return <p>Mock Power Selection</p>;
    }
);

describe('App component', () => {
  let testContext: {
    component?: renderer.ReactTestRenderer;
  };

  const renderComponent = () => <App />;

  beforeEach(() => {
    testContext = {};
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('triggers calls for initial fetching of data when rendered', async () => {
    const dispatch = jest.fn();
    (useAppDispatch as jest.Mock).mockReturnValue(dispatch);

    await renderer.act(async () => {
      testContext.component = renderer.create(renderComponent());
    });

    expect(dispatch.mock.calls).toEqual([
      ['MockFetchCharactersAction'],
      ['MockFetchPowersAction'],
    ]);
  });

  it('renders a header area and top-level routes', async () => {
    const dispatch = jest.fn();
    (useAppDispatch as jest.Mock).mockReturnValue(dispatch);

    await renderer.act(async () => {
      testContext.component = renderer.create(renderComponent());
    });

    expect(testContext.component).toMatchSnapshot();
  });

  it('aborts thunks when unmounted', async () => {
    const charactersDispatchMock = jest.fn();
    const powersDispatchMock = jest.fn();

    const dispatch = jest
      .fn()
      .mockReturnValueOnce({ abort: charactersDispatchMock })
      .mockReturnValueOnce({ abort: powersDispatchMock });

    (useAppDispatch as jest.Mock).mockReturnValue(dispatch);

    await renderer.act(async () => {
      testContext.component = renderer.create(renderComponent());
    });

    expect(dispatch.mock.calls).toEqual([
      ['MockFetchCharactersAction'],
      ['MockFetchPowersAction'],
    ]);

    await renderer.act(async () => {
      testContext.component?.unmount();
    });

    expect(charactersDispatchMock).toHaveBeenCalledTimes(1);
    expect(powersDispatchMock).toHaveBeenCalledTimes(1);
  });
});
