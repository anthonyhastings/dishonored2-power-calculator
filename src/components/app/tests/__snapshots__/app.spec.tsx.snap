// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component when rendered renders a header area and top-level routes 1`] = `
<div
  className="app"
>
  <header
    className="app__header"
  >
    <h1
      className="app__title"
    >
      Power Calculator
    </h1>
    <MockLink
      className="app__logo-link"
      to="/"
    >
      <img
        alt="Dishonored 2 logo"
        className="app__logo-graphic"
        draggable="false"
        height="90"
        sizes="100vw"
        srcSet="660x90.png 660w, 1320x180.png 1320w"
        width="660"
      />
    </MockLink>
  </header>
  <div
    className="app__content"
  >
    <MockLoader
      showError={false}
      showLoader={true}
    >
      <MockRoutes>
        <span
          path="*"
        >
          <p>
            Page Not Found
          </p>
        </span>
        <span
          path="/"
        >
          <p>
            Mock Character Selection
          </p>
        </span>
        <span
          path="/powers/:characterSlug"
        >
          <p>
            Mock Power Selection
          </p>
        </span>
      </MockRoutes>
    </MockLoader>
  </div>
</div>
`;
